### PR TITLE
Add metadata to Hearth::Output and Errors, and return Hearth::Output on successful calls

### PR DESCRIPTION
**Pull request by mullermp**
Breaking change.

Return Hearth::Output from operations, instead of the output structure. This is similar to V3 (without delegation however).

This allows middleware to set context values and have it be accessible to users.

### CODE REVIEW

**Comment by mullermp** [comment]
WIP - needs update on tests and potentially breaking CI (it SHOULD break?)

**Review by alextwoods — COMMENTED** [review]
Looks good overall.  We should add an rbs for Output now thats its public.

File: hearth/lib/hearth/http/api_error.rb [hunk 1]
```diff
@@ -9,7 +9,6 @@ def initialize(http_resp:, **kwargs)
         @http_status = http_resp.status
         @http_headers = http_resp.headers
         @http_body = http_resp.body
-        @request_id = http_resp.headers['x-request-id']
```
Comment (mullermp) [inline hearth/lib/hearth/http/api_error.rb:12]: This is considered "common" and not "standard" in http spec, so I opted to remove it here.

File: codegen/smithy-ruby-codegen/src/main/java/software/amazon/smithy/ruby/codegen/middleware/MiddlewareStackStep.java [hunk 2]
```diff
@@ -42,8 +42,8 @@ public String toString() {
                 return "Retry";
             case SERIALIZE:
                 return "Serialize";
-            case BUILD:
-                return "Build";
+            case VALIDATE:
```
Comment (mullermp) [inline codegen/smithy-ruby-codegen/src/main/java/software/amazon/smithy/ruby/codegen/middleware/MiddlewareStackStep.java:45]: Build is confusing with serialize. I think we needed one at the validate step.

File: codegen/smithy-ruby-rails-codegen/middleware/request_id.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+module Middleware
```
Comment (mullermp) [inline codegen/smithy-ruby-rails-codegen/middleware/request_id.rb:1]: This might be global scope isn't it? I think we'd need to write out Service::Middleware

File: codegen/smithy-ruby-rails-codegen/middleware/request_id.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  # on the context.
+  # @api private
+  class RequestId
+    def initialize(app)
+      @app = app
+    end
+
+    def call(input, context)
+      output = @app.call(input, context)
+      output.metadata[:request_id] = context.response.headers['x-request-id']
```
Comment (mullermp) [inline codegen/smithy-ruby-rails-codegen/middleware/request_id.rb:12]: I don't like that I had to parse here (for data) and in error parser (for errors). I'd like one place to put this information.

File: codegen/smithy-ruby-rails-codegen/middleware/request_id.rb (hunk 3 above)
Comment (alextwoods) [reply codegen/smithy-ruby-rails-codegen/middleware/request_id.rb:1]: Yeah, this either needs to be in a core gem or needs to be codegened out with a unique namespace....

File: codegen/smithy-ruby-codegen/src/main/java/software/amazon/smithy/ruby/codegen/ApplicationTransport.java [hunk 5]
```diff
@@ -126,7 +126,7 @@ public static ApplicationTransport createDefaultHttpApplicationTransport() {
             List<Middleware> middleware = new ArrayList();
             middleware.add((new Middleware.Builder())
                     .klass("Hearth::HTTP::Middleware::ContentLength")
-                    .step(MiddlewareStackStep.BUILD)
+                    .step(MiddlewareStackStep.SERIALIZE)
```
Comment (alextwoods) [inline codegen/smithy-ruby-codegen/src/main/java/software/amazon/smithy/ruby/codegen/ApplicationTransport.java:129]: Why is this moved to Serialize?

File: codegen/smithy-ruby-codegen/src/main/java/software/amazon/smithy/ruby/codegen/middleware/MiddlewareStackStep.java (hunk 2 above)
Comment (alextwoods) [reply codegen/smithy-ruby-codegen/src/main/java/software/amazon/smithy/ruby/codegen/middleware/MiddlewareStackStep.java:45]: These steps were intended to match those used by other SDKs. See for example
* https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/modules/_aws_sdk_middleware_stack.html
* https://aws.github.io/aws-sdk-go-v2/docs/middleware/#:~:text=You%20can%20customize%20AWS%20SDK,step's%20input%20and%20output%20types.

The meanings were intended to be:
* Serialize - Serialize the given parameter input into a request data structure that can be confusmed by the transport layer (eg, an HTTP:Request).
* Build - Provide additional transformation of the Request after serialization (eg ContentLength)

I agree however that because of our choice to use "Builders" for Serialization that this is a bit confusing.  I don't think VALIDATE is the right name for this step though.

File: codegen/smithy-ruby-rails-codegen/middleware/request_id.rb (hunk 4 above)
Comment (alextwoods) [reply codegen/smithy-ruby-rails-codegen/middleware/request_id.rb:12]: I agree - I'd rather be able to do this in a single place.  But I don't believe its possible without some interface tweak, either adding output to the error parser interface (to allow access to metadata) or to add write access to the request_id.

We can add either of those later if we want (but taking them away would be hard).  So, lets leave this for now and accept some duplication.

**Review by alextwoods — COMMENTED** [review]
Overall looks good and I think we should move forward with changing our return type.

File: codegen/smithy-ruby-codegen/src/main/java/software/amazon/smithy/ruby/codegen/ApplicationTransport.java (hunk 5 above)
Comment (mullermp) [reply codegen/smithy-ruby-codegen/src/main/java/software/amazon/smithy/ruby/codegen/ApplicationTransport.java:129]: I removed Build because it didn't make sense, it's now "validate". Content length should be after serialize.

File: codegen/smithy-ruby-codegen/src/main/java/software/amazon/smithy/ruby/codegen/middleware/MiddlewareStackStep.java (hunk 2 above)
Comment (mullermp) [reply codegen/smithy-ruby-codegen/src/main/java/software/amazon/smithy/ruby/codegen/middleware/MiddlewareStackStep.java]: I just undid it. Can re-evaluate later.

**Review by alextwoods — APPROVED** [review]

File: codegen/smithy-ruby-rails-codegen/middleware/request_id.rb (hunk 4 above)
Comment (mullermp) [reply codegen/smithy-ruby-rails-codegen/middleware/request_id.rb:12]: Talked offline, metadata is now a property of errors.

File: codegen/smithy-ruby-codegen-test/integration-specs/errors_spec.rb [hunk 6]
```diff
@@ -10,6 +10,17 @@ module Errors
         error = ApiError.new(http_resp: http_resp, error_code: 'error')
         expect(error).to be_a(Hearth::HTTP::ApiError)
       end
+
+      it 'stores a request_id' do
```
Comment (alextwoods) [inline codegen/smithy-ruby-codegen-test/integration-specs/errors_spec.rb]: Does this still apply or should it be metadata?

File: codegen/smithy-ruby-codegen/src/main/java/software/amazon/smithy/ruby/codegen/generators/ErrorsGeneratorBase.java [hunk 7]
```diff
@@ -141,13 +144,13 @@ private void renderRbsBaseErrors() {
     }
 
     /**
-     * Called to render the error_code method.
+     * Called to render the error_code method body.
```
Comment (alextwoods) [inline codegen/smithy-ruby-codegen/src/main/java/software/amazon/smithy/ruby/codegen/generators/ErrorsGeneratorBase.java:147]: Unrelated nit: We should probably describe what error code is (ie, the name of the class to use....  Because this function also may map the actual data to a generated class, including reserved word name changes...).

File: hearth/lib/hearth/http/error_parser.rb [hunk 8]
```diff
@@ -30,22 +30,17 @@ class ErrorParser
       #
       # @param [Array<Class<ApiError>>] errors Array of Error classes
       #   modeled for the operation.
-      #
-      # @param [callable] error_code_fn Protocol specific function
-      #   that will return the error code from a response, or nil if
-      #   there is none.
-      def initialize(error_module:, success_status:, errors:, error_code_fn:)
+      def initialize(error_module:, success_status:, errors:)
```
Comment (alextwoods) [inline hearth/lib/hearth/http/error_parser.rb:33]: Nit: add to documenation of error_module, it must contain definition of error code function

File: hearth/lib/hearth/http/error_parser.rb [hunk 9]
```diff
@@ -30,22 +30,17 @@ class ErrorParser
...
         @error_module = error_module
         @success_status = success_status
         @errors = errors
-        @error_code_fn = error_code_fn
       end
 
       # Parse and return the error if the response is not successful.
       #
       # @param [Response] response The HTTP response
```
Comment (alextwoods) [inline hearth/lib/hearth/http/error_parser.rb:41]: Nit: Update @params for metadata

**Review by alextwoods — APPROVED** [review]

File: codegen/smithy-ruby-codegen-test/integration-specs/errors_spec.rb (hunk 6 above)
Comment (mullermp) [reply codegen/smithy-ruby-codegen-test/integration-specs/errors_spec.rb]: Fixed and deleted
